call_option.go

```diff
@@ -1,118 +1,116 @@
+//     * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
 package gax
 
 import (
+	"math/rand"
 	"time"
 
+	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 )
 
 type CallOption interface {
 	Resolve(*CallSettings)
 }
 
-type callOptions []CallOption
+type retryerOption func() Retryer
+
+func (o retryerOption) Resolve(s *CallSettings) {
+	s.Retryer = o
+}
 
-func (opts callOptions) Resolve(s *CallSettings) *CallSettings {
-	for _, opt := range opts {
-		opt.Resolve(s)
+func WithBackoffRetryer(base, max time.Duration, mult float64, codes ...codes.Code) CallOption {
+	if len(codes) == 0 {
+		return retryerOption(nil)
 	}
-	return s
+	return retryerOption(func() Retryer {
+		return &codeRetryer{
+			b: &exponentialBackoff{Base: base, Max: max, Mult: mult},
+			c: codes,
+		}
+	})
 }
 
-// Encapsulates the call settings for a particular API call.
 type CallSettings struct {
-	Timeout       time.Duration
-	RetrySettings RetrySettings
+	// Retryer returns a Retryer to be used to control retry logic of a method call.
+	// If retry is undesirable, simply set the function to nil.
+	// If the function is not nil, it must return non-nil Retryer.
+	Retryer func() Retryer
 }
 
-// Per-call configurable settings for retrying upon transient failure.
-type RetrySettings struct {
-	RetryCodes      map[codes.Code]bool
-	BackoffSettings BackoffSettings
+type Retryer interface {
+	// Retry decides whether a request should be retried and how long to wait before retrying
+	// if the previous attempt returned with err. err is never nil.
+	Retry(err error) (backoff time.Duration, ok bool)
 }
 
-// Parameters to the exponential backoff algorithm for retrying.
-type BackoffSettings struct {
-	DelayTimeoutSettings MultipliableDuration
-	RPCTimeoutSettings   MultipliableDuration
+type BackoffStrategy interface {
```

Comment:
I don't think we need to expose this interface. We have Retryer for the general case, and we have an implementation of exponential backoff because it's hard to write. That's all the user needs.